### PR TITLE
AK + Kernel: Make AsyncDeviceRequest sub-req management alloc free with AK::IntrusiveList

### PR DESCRIPTION
**Pull request by bgianfo**
-  __AK: Add some initial unit tests for AK::IntrusiveList__

-   __AK: Fix IntrusiveList::take_first/last() actually compile with RefPtr<T>__

    PR #6376 made IntrusiveList capable of holding RefPtr<T>, etc. however
    there was a latent bug where take_first() / take_last() would fail to
    compile because they weren't being converted to their container type.

-   __AK: Add unit tests to verify AK::InstrusiveList<T> holding RefPtr<T>__

    Validate that the refcounting works as expected, and continues to work.

-   __Kernel: Make AsyncDeviceRequest sub-req management alloc free__

    The previous implementation could allocate on insertion into the completed / pending
    sub request vectors. There's no reason these can't be intrusive lists instead.

    This is a very minor step towards improving the ability to handle OOM, as tracked by #6369
    It might also help improve performance on the IO path in certain situations.
    I'll benchmark that later.

-   __Kernel: Annotate more AsyncDeviceRequest API's with [[nodiscard]]__

CC: @alimpfard for the AK::IntrusiveList<T> fixes.

### CODE REVIEW

File: AK/Tests/TestIntrusiveList.cpp [hunk 1]
```diff
@@ -73,4 +71,78 @@ TEST_CASE(enumeration)
...
+        IntrusiveRefPtrList ref_list;
+        ref_list.append(*item);
+        EXPECT_EQ(2u, item->ref_count());
+    }
+
+    EXPECT_EQ(1u, item->ref_count());
+}
+
+// There is currently a bug in the NonnullRefPtr specialization of IntrusiveList,
+// so this test fails to compile at the moment. Including it for future use.
```
Comment (alimpfard) [inline AK/Tests/TestIntrusiveList.cpp:122]: :thinking:

How does this fail?
I do remember checking that it compiles, though not with a lot of method calls...

Comment (bgianfo) [reply AK/Tests/TestIntrusiveList.cpp:122]: Maybe I'm holding it wrong? 

```
inja: Entering directory `/home/bgianf/src/serenity/Build/i686'
[0/2] Re-checking globbed directories...
[2/3] Building CXX object AK/Tests/CMakeFiles/TestIntrusiveList.dir/TestIntrusiveList.cpp.o
FAILED: AK/Tests/CMakeFiles/TestIntrusiveList.dir/TestIntrusiveList.cpp.o
/usr/bin/ccache /home/bgianf/src/serenity/Toolchain/Local/i686/bin/i686-pc-serenity-g++ -DDEBUG -DENABLE_COMPILETIME_FORMAT_CHECK -DSANITIZE_PTRS -I../../Userland/Libraries -I../../. -I. -I../../Userland/Libraries/LibC -I../../Userland/Libraries/LibM -I../../Userland/Libraries/LibSystem -I../../Userland/Services -I../../Userland -IUserland/Services -IUserland/Libraries -IUserp
In file included from ../../AK/Tests/TestIntrusiveList.cpp:27:
../.././AK/IntrusiveList.h: In instantiation of 'class AK::IntrusiveList<IntrusiveNonnullRefPtrItem, AK::NonnullRefPtr<IntrusiveNonnullRefPtrItem>, &IntrusiveNonnullRefPtrItem::m_list_node>':
../../AK/Tests/TestIntrusiveList.cpp:135:32:   required from here
../.././AK/IntrusiveList.h:352:7: error: could not convert template argument '&IntrusiveNonnullRefPtrItem::m_list_node' from 'AK::IntrusiveListNode<IntrusiveNonnullRefPtrItem, AK::NonnullRefPtr<IntrusiveNonnullRefPtrItem> > IntrusiveNonnullRefPtrItem::*' to 'AK::IntrusiveListNode<IntrusiveNonnullRefPtrItem, AK::RefPtr<IntrusiveNonnullRefPtrItem, AK::RefPtrTraits<IntrusiveNonn'
  352 | class IntrusiveList<T, NonnullRefPtr<T>, member> : public IntrusiveList<T, RefPtr<T>, member> {
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../AK/Tests/TestIntrusiveList.cpp: In function 'void __test_intrusive_nonnull_ref_ptr_intrusive()':
../../AK/Tests/TestIntrusiveList.cpp:137:22: error: 'using IntrusiveNonnullRefPtrList = class AK::IntrusiveList<IntrusiveNonnullRefPtrItem, AK::NonnullRefPtr<IntrusiveNonnullRefPtrItem>, &IntrusiveNonnullRefPtrItem::m_list_node>' {aka 'class AK::IntrusiveList<IntrusiveNonnullRefPtrItem, AK::NonnullRefPtr<IntrusiveNonnullRefPtrItem>, &IntrusiveNonnullRefPtrItem::m_list_node>'}'
  137 |     nonnull_ref_list.append(*item);
```

Comment (alimpfard) [reply AK/Tests/TestIntrusiveList.cpp:122]: Ohhh it needs to entirely rewrite `IntrusiveListNode<T, NonnullRefPtr<T>>` to `IntrusiveListNode<T, RefPtr<T>>`, not just subclass it
I guess I made the node hold a `RefPtr<T>` when I tested it...

It'll get a little ugly, but should be fixable...

Comment (alimpfard) [reply AK/Tests/TestIntrusiveList.cpp:122]: I made a PR on your repo to this branch with a fix :^)

Comment (alimpfard) [reply AK/Tests/TestIntrusiveList.cpp:122]: ...which apparently doesn't fix anything...

Comment (bgianfo) [reply AK/Tests/TestIntrusiveList.cpp:122]: :) Did you try un-ifdefing that test?

Comment (alimpfard) [reply AK/Tests/TestIntrusiveList.cpp:122]: hmm...weird, it works for me, I wonder if it's a clang thing?

Comment (alimpfard) [reply AK/Tests/TestIntrusiveList.cpp:122]: I `#ifdef`'d around clang, hopefully it builds and you can un-if0 your other test.

Comment (bgianfo) [reply AK/Tests/TestIntrusiveList.cpp:122]: Do you  just want to send this as a separate PR?

I don't need NonnullRefPtr here for any reason, just wanted to add a test to cover it.
This shouldn't be gated on that working?

Comment (alimpfard) [reply AK/Tests/TestIntrusiveList.cpp:122]: Either way is fine by me, I can add this fix and un-if0 the test (or just add it) in a new PR.

Comment (bgianfo) [reply AK/Tests/TestIntrusiveList.cpp:122]: I would prefer that, thanks!

**Comment by bgianfo** [comment]
@awesomekling / @linusg  any feedback?

**Comment by tomuta** [comment]
Looks good to me and makes sense.

**Pull request merged**